scripts/tosback-import.js

```diff
@@ -0,0 +1,155 @@
+  return path.join(LOCAL_TOSBACK2_REPO, TOSBACK2_RULES_FOLDER_NAME);
+}
+
+function getGitHubWebUrl(commitHash, filename) {
+  return [
+    TOSBACK2_WEB_ROOT,
+    'blob',
+    commitHash,
+    TOSBACK2_RULES_FOLDER_NAME,
+    filename
+  ].join('/');
+}
+
+async function parseFile(filename) {
+  const data = await fs.readFile(filename);
+  return parser.toJson(data);
+}
+
+function toPascalCase(str) {
+  const lowerCase = str.toLowerCase();
+  return str[0].toUpperCase() + lowerCase.substring(1);
+}
+
+function toType(str) {
+  let found;
+  for (const i in TYPES) {
+    if (TYPES[i].name === str) {
+      found = i;
+      break;
+    }
+  }
+  if (!found) {
+    throw new Error(`Unsupported type: ${str}`);
+  }
+  return found;
+}
+
+async function getFiltered(docObj) {
+  const content = await fetch(docObj.fetch);
+  return filter(content, docObj, []);
+}
+
+async function validDocObj(docObj) {
+  try {
+    const filtered = await getFiltered(docObj);
+    const filtered2 = await getFiltered(docObj);
+    if (filtered !== filtered2) {
+      return false;
+    }
+    return (filtered.length > 100);
+  } catch (e) {
+    console.log('validation error', e);
+    return false;
+  }
+}
+
+async function parseTosback2(importedFrom, imported) {
+  if (!Array.isArray(imported.sitename.docname)) {
+    imported.sitename.docname = [imported.sitename.docname];
+  }
+  const siteName = toPascalCase(imported.sitename.name.split('.')[0]);
+  const documents = {};
+  const promises = imported.sitename.docname.map(async docnameObj => {
+    const type = toType(docnameObj.name);
+    if (documents.type) {
+      throw new Error('Same type used twice!');
+    }
+    const docObj = {
+      fetch: docnameObj.url.name
+    };
+
+    try {
+      docObj.select = xPathToCss(docnameObj.url.xpath);
+    } catch (e) {
+      docObj.select = 'body';
+      if (docnameObj.url.xpath) {
+        console.error('XPath-to-CSS failed:', docnameObj.url.xpath);
+      }
+    }
+    const valid = await validDocObj(docObj);
+    if (valid) {
+      documents[type] = docObj;
+    }
+  });
+  await Promise.all(promises);
+  const declaration = {
+    importedFrom,
+    name: siteName,
+    documents
+  };
+  try {
+    assertValid(serviceSchema, declaration);
+  } catch (e) {
+    console.log(e);
+    throw e;
+  }
+  return declaration;
+}
+
+async function parseAll(folder) {
+  const git = simpleGit(folder);
+  const gitLog = await git.log();
+  const commitHash = gitLog.latest.hash;
+
+  const files = await fs.readdir(folder);
+  const promises = files.map(async filename => {
+    // Uncomment this to run the script on a single XML file:
+    // if (filename !== 'zoosk.com.xml') return;
+    let imported;
+    try {
+      imported = JSON.parse(await parseFile(path.join(folder, filename)));
+    } catch (e) {
+      console.error('Error parsing xml', filename, e.message);
+      throw e;
+    }
+    console.log('imported', filename, JSON.stringify(imported, null, 2));
```

Comment:
Let me know if you want me to remove the console.log statements from the script.

Reply:
At this stage we also log directly from the code. I'd like to move that into an events-based logger at some stage, but right now this is fine.
Also, as you mentioned above, importing from tosback is definitely a dev thing, not an end-user feature, so I feel it is fine to expose a potentially verbose CLI as long as it helps understanding progress, success and failure (as opposed to debugging information)  🙂